wgengine/ifstatus_windows.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package wgengine
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/tailscale/wireguard-go/tun"
+	"golang.zx2c4.com/wireguard/windows/tunnel/winipcfg"
+	"tailscale.com/types/logger"
+)
+
+type ifaceWatcher struct {
+	mu   sync.Mutex
+	logf logger.Logf
+	luid winipcfg.LUID
+	done bool
+	sig  chan bool
+}
+
+func (iw *ifaceWatcher) callback(notificationType winipcfg.MibNotificationType, iface *winipcfg.MibIPInterfaceRow) {
+	// Probably should check only when MibParameterNotification, but just in case included MibAddInstance also.
+	if notificationType == winipcfg.MibParameterNotification || notificationType == winipcfg.MibAddInstance {
+		go iw.isUp()
```

Comment:
Why do you have to use `go` here? Comment would help.

Reply:
I always use `go` for OS events / callbacks, to avoid keeping their threads busy for any amount of time, even if it's short. Not sure that it's needed though. I simply always did it this way. Do you want me to change or to comment?

Reply:
In this particular case `iw.isUp()` internally does another Windows SDK call, and this is especially what I want to avoid.

Reply:
I was surprised to see goroutine started without any synchronisation code near by - `ifaceWatcher.callback` does pretty much nothing else but starts goroutine. So any comment that explains the reason is welcome. Thank you.

I am not sure about `keeping OS thread busy` part. Goroutines are automatically scheduled on many different OS threads, so it is quite possible that both `ifaceWatcher.callback` and `iw.isUp` are running on the same thread. But it is fine to leave the comment, if you think it is appropriate.

Also it is quite legal (from Go perspective) to call Windows API from Windows callback. It is certainly more rare than calling Windows API from normal goroutine. But starting goroutine from Windows callback is also rare.

Maybe what is important is that Windows callback returns as quickly as possible. So it can be called again by Windows without prior callback still not returned.

I will let you and Brad decide what to do here.

Reply:
> I was surprised to see goroutine started without any synchronisation code near by

Yep, bit strange looking, but from the callback's point of view it's simply _fire-and-forget_ thing, no need for any synchronization.

Regarding if async is needed or not in this particular case - I don't know - will do how ever you guys decide. But what I know for sure is that there are some restrictions with some Windows API callbacks, no matter the language, like you should not **unregister the callback from the callback** (maybe it internally enumerates callbacks to execute them, and removing an item breaks enumeration or something - not sure). I've read that somewhere for some callback, but can't remember where. For such reasons I simply always runaway from the callback asap, and then doing my thing.

Again, I don't claim that this is the fact, and that it has to be done async - just explaining my reasons. 

Is there any reason to avoid `go` that I should be aware of?

Reply:
I don't know of a reason to avoid `go` in a Windows callback.

Your last version of this comment text SGTM. "Out of paranoia" sums up our feeling.

Thank you,